### PR TITLE
smartcard: implement ListReaderGroups

### PR DESCRIPTION
**Pull request by RoadRunnr**

### CODE REVIEW

**Comment by bmiklautz** [comment]
@freerdp-bot test

File: channels/smartcard/client/smartcard_operations.c [hunk 1]
```diff
@@ -1246,11 +1335,13 @@ UINT32 smartcard_irp_device_control_decode(SMARTCARD_DEVICE* smartcard, SMARTCAR
 			break;
 
 		case SCARD_IOCTL_LISTREADERGROUPSA:
-			status = SCARD_F_INTERNAL_ERROR;
+			call = calloc(1, sizeof(ListReaderGroups_Call));
```
Comment (bmiklautz) [inline channels/smartcard/client/smartcard_operations.c]: unchecked calloc.

File: channels/smartcard/client/smartcard_operations.c [hunk 2]
```diff
@@ -1246,11 +1335,13 @@ UINT32 smartcard_irp_device_control_decode(SMARTCARD_DEVICE* smartcard, SMARTCAR
...
+			status = smartcard_ListReaderGroupsA_Decode(smartcard, operation, (ListReaderGroups_Call*) call);
 			break;
 
 		case SCARD_IOCTL_LISTREADERGROUPSW:
-			status = SCARD_F_INTERNAL_ERROR;
+			call = calloc(1, sizeof(ListReaderGroups_Call));
```
Comment (bmiklautz) [inline channels/smartcard/client/smartcard_operations.c]: unchecked calloc

**Comment by bmiklautz** [comment]
I saw that smartcard_ListReaderGroupsX_Decode returns with an memory error in case the  calloc failed. However I'd suggest to directly check the calloc and return an different error in case call isn't set within the function.

File: channels/smartcard/client/smartcard_pack.c [hunk 3]
```diff
@@ -534,6 +534,130 @@ void smartcard_trace_long_return(SMARTCARD_DEVICE* smartcard, Long_Return* ret,
...
+
+	if (unicode)
+	{
+		length = ret->cBytes / 2;
+		ConvertFromUnicode(CP_UTF8, 0, (WCHAR*) ret->msz, length, &mszA, 0, NULL, NULL);
+	}
+	else
+	{
+		length = ret->cBytes;
+		mszA = (char*) malloc(length);
```
Comment (bmiklautz) [inline channels/smartcard/client/smartcard_pack.c]: unchecked malloc

Comment (RoadRunnr) [inline channels/smartcard/client/smartcard_pack.c]: I copied that from from one of the other trace functions. It seems someone thought the trace functions do not need this kind protection.

I'll push another cset on top of this that addresses *all* mallocs in trace returns.

Comment (bmiklautz) [inline channels/smartcard/client/smartcard_pack.c]: @RoadRunnr there is some ongoing hardening work for existing code that should over the open unchecked functions but new code that is added should handle it properly already.

**Comment by RoadRunnr** [comment]
> I saw that smartcard_ListReaderGroupsX_Decode returns with an memory error in case the calloc failed. However I'd suggest to directly check the calloc and return an different error in case call isn't set within the function.

I only followed the style of the other functions. But looking at the actual code I see a potential for a cleanup there. I'll cook something up.

**Comment by bmiklautz** [comment]
@RoadRunnr thank you very much. But as said the already existing code is actively worked on and gets return value fixes and such. I'd recommend that you add the checks only in your pull request for now (then a merge is easier) and if you in mood you could check and review the code after the pending work is pushed.

**Comment by RoadRunnr** [comment]
@bmiklautz uhm, already did the changes before I read your comments.

Do you want me to drop the additional changes?

**Comment by bmiklautz** [comment]
@RoadRunnr nope it's fine. We will handle the merge than. Thanks.

**Comment by akallabeth** [comment]
@bmiklautz Looks like we forgot about this one. Should rebase and merge if possible.

**Pull request closed without merging**